### PR TITLE
Defer adding hooks until the widgets are registered

### LINKED ISSUE #98: Ensure widget is still registered at running hook
Since hooks are added in the constructor (not ideal) when the hook runs it should ideally check to make sure the widget is still enqueued.

### PR DESCRIPTION
**Pull request by westonruter**
Fixes #98.

### CODE REVIEW

**Comment by obenland** [comment]
So you're saying let the unregistering party deal with cleanup?
(Which I'm absolutely fine with, just wanted to clarify)

**Comment by westonruter** [comment]
@obenland no, this approach which you suggested will avoid the need to cleanup. if `unregister_widget()` is called before widgets are registered, as it would be, then the `_register` method would never get called on the unregistered widget class and there would be nothing to clean up.

**Comment by obenland** [comment]
Oh, right, core widgets don't get registered until priority 100.

**Comment by westonruter** [comment]
To clarify, as I see there is some overloading of the term “register”: when you call `register_widget()` it doesn't actually call `WP_Widget::_register()`. The `register_widget()` call happens at `widgets_init` priority 10, and then at `widgets_init` priority 100 when `WP_Widget_Factory::_register_widgets()` is called, and this factory method then actually iterates over all of the registered widgets (classes) and then registers all of the widget instances.

File: wp-includes/widgets/class-wp-widget-media.php [hunk 1]
```diff
@@ -78,6 +78,15 @@ public function __construct( $id_base, $name, $widget_options = array(), $contro
 			$widget_opts,
 			$control_opts
 		);
+	}
+
+	/**
+	 * Add hooks while registering all widget instances of this widget class.
+	 *
+	 * @since 2.8.0
```
Comment (obenland) [inline wp-includes/widgets/class-wp-widget-media.php]: Docs typo

**Review by obenland — APPROVED** [review]
One docs typo, looks good otherwise

**Pull request merged**